File: cpp/cpp-boost/demo18a02-barrier.cpp
/*
BARRIERS
Version A: Cyclic barriers
*/


#include <iostream>
#include <string>
#include <boost/tuple/tuple.hpp>
#include <boost/chrono.hpp>
#include <boost/thread.hpp>
using namespace std;



typedef boost::tuple<string,int> tuplestrint;

boost::barrier syncPoint(2); // participant count = 2



void processRequest(string userName, int timeWait) {
    boost::this_thread::sleep_for(boost::chrono::seconds(timeWait));

    cout << "Get request from " << userName << endl;

    syncPoint.count_down_and_wait();

    cout << "Process request for " << userName << endl;

    syncPoint.count_down_and_wait();

    cout << "Done " << userName << endl;
}



int main() {
    const int NUM_THREADS = 4;
    boost::thread_group lstTh;

    // tuple<userName, timeWait>
    tuplestrint lstArg[NUM_THREADS] = {
        tuplestrint("lorem", 1),
        tuplestrint("ipsum", 3),
        tuplestrint("dolor", 3),
        tuplestrint("amet", 10),
    };

    for (int i = 0; i < NUM_THREADS; ++i) {
        tuplestrint & arg = lstArg[i];
        lstTh.add_thread(new boost::thread(
            &processRequest, boost::get<0>(arg), boost::get<1>(arg)
        ));
    }

    // Thread with userName = "amet" shall be FREEZED

    lstTh.join_all();

    return 0;
}

File: python/demo18a02_barrier.py
'''
BARRIERS AND LATCHES
Version A: Barriers
'''

import time
import threading



sync_point = threading.Barrier(parties=2)



def process_request(user_name: str, time_wait: int):
    time.sleep(time_wait)

    print(f'Get request from {user_name}')
    sync_point.wait()

    print(f'Process request for {user_name}')
    sync_point.wait()

    print(f'Done {user_name}')



lstarg = [
    ('lorem', 1),
    ('ipsum', 3),
    ('dolor', 3),
    ('amet', 10)
]

_ = [ threading.Thread(target=process_request, args=arg) for arg in lstarg ]

# Thread with user_name = "amet" shall be FREEZED
